### PR TITLE
Add support for style=form explode=false to serialized parameter

### PR DESCRIPTION
**Pull request by ctipggroup**
Add support for style=form explode=false to serialized parameter

### CODE REVIEW

**Comment by ctipggroup** [comment]
This is add support for

`
- in: query
          name: id
          style: form
          explode: false
          schema:
            type: array
            items:
              type: int
`

?id=1,2,3

**Comment by scaytrase** [comment]
Can you please provide new test verifying that new validation works (both successful and failing cases)

**Comment by ctipggroup** [comment]
> Can you please provide new test verifying that new validation works (both successful and failing cases)

Updated with tests. 
Thanks

File: src/PSR7/Validators/SerializedParameter.php [hunk 1]
```diff
@@ -85,30 +94,47 @@ public function deserialize($value)
...
+
+    private function isJsonContentType() : bool
+    {
+        return $this->contentType !== null && preg_match('#^application/.*json$#', $this->contentType) !== false;
+    }
+
+    private function convertScalar($value, string $type)
+    {
+        if (($type === CebeType::BOOLEAN) && is_scalar($value) && preg_match('#^(true|false)$#i', (string) $value)) {
+            return strtolower($value) === 'true';
```
Comment (ctipggroup) [inline src/PSR7/Validators/SerializedParameter.php:120]: Fixed to return false for 'false'

**Comment by ctipggroup** [comment]
> Can you please provide new test verifying that new validation works (both successful and failing cases)

Tests added
Added support to convert array items to correct item type
Fixed issue where 'false' is evaluated to true

File: src/PSR7/Validators/SerializedParameter.php [hunk 2]
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
...
+    {
+        return $this->contentType !== null && preg_match('#^application/.*json$#', $this->contentType) !== false;
+    }
+
+    /**
+     * @param mixed $value
+     *
+     * @return mixed
+     */
+    private function convertScalar($value, ?string $type = '')
```
[the remark this answers is not in the document]
Comment (ctipggroup) [reply src/PSR7/Validators/SerializedParameter.php]: Removed

File: src/PSR7/Validators/SerializedParameter.php [hunk 3]
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
             return $decodedValue;
         }
 
-        if (($this->schema->type === CebeType::BOOLEAN) && is_scalar($value) && preg_match('#^(true|false)$#i', (string) $value)) {
-            return (bool) $value;
+        $value = $this->convertScalar($value, $this->schema->type);
```
[the remark this answers is not in the document]
Comment (ctipggroup) [reply src/PSR7/Validators/SerializedParameter.php]: Refactored as suggested.

File: src/PSR7/Validators/SerializedParameter.php [hunk 4]
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
...
+
+        if (($this->schema->type === CebeType::ARRAY)
+            && ($this->style === 'form' && $this->explode === false)
+            && is_string($value)) {
+            $value = explode(',', $value);
+            foreach ($value as &$val) {
+                $val = $this->convertScalar($val, $this->schema->items->type ?? '');
```
[the remark this answers is not in the document]
Comment (ctipggroup) [reply src/PSR7/Validators/SerializedParameter.php]: Changed to null as suggested

File: src/PSR7/Validators/SerializedParameter.php [hunk 5]
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
             return $decodedValue;
         }
 
-        if (($this->schema->type === CebeType::BOOLEAN) && is_scalar($value) && preg_match('#^(true|false)$#i', (string) $value)) {
-            return (bool) $value;
+        $value = $this->convertScalar($value, $this->schema->type);
+
+        if (($this->schema->type === CebeType::ARRAY)
+            && ($this->style === 'form' && $this->explode === false)
```
[the remark this answers is not in the document]
Comment (ctipggroup) [reply src/PSR7/Validators/SerializedParameter.php]: Refactored as sugggested

[the remark this answers is not in the document]
Comment (ctipggroup) [reply src/PSR7/Validators/SerializedParameter.php]: Constants added

File: tests/FromCommunity/IssueWithQueryArrayTest.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        '200':
+          description: A list of users
+YAML;
+    }
+
+    protected function makeRequest(string $style, string $type) : ServerRequest
+    {
+        $map     = [
+            'form' => ['integer' => '1,2,3', 'string' => 'id1,id2,id3', 'boolean' => 'true,false', 'number' => '1.00,2.00,3.00'],
+            'spaceDelimited' => ['integer' => '1%202%203', 'string' => 'id1%20id2%20id3', 'boolean' => 'true%20false', 'number' => '1.00%202.00%203.00'],
```
Comment (scaytrase) [inline tests/FromCommunity/IssueWithQueryArrayTest.php:104]: I'm not sure whether real world requests have encoded values in request object

Comment (ctipggroup) [reply tests/FromCommunity/IssueWithQueryArrayTest.php]: Updated

**Comment by scaytrase** [comment]
Great work! Thanks @ctipggroup

**Pull request merged**